In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("Pharmacy_OLTP_SLStyle_18Months.csv")

df.head()
df.shape
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 27 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Transaction_ID         50000 non-null  int64  
 1   Transaction_Date       50000 non-null  object 
 2   Branch_ID              50000 non-null  int64  
 3   Branch_Name            50000 non-null  object 
 4   District               50000 non-null  object 
 5   Province               50000 non-null  object 
 6   Payment_Method         50000 non-null  object 
 7   Customer_Gender        50000 non-null  object 
 8   Customer_Age           50000 non-null  int64  
 9   Customer_City          50000 non-null  object 
 10  Medicine_ID            50000 non-null  int64  
 11  Medicine_Name          50000 non-null  object 
 12  Generic_Name           50000 non-null  object 
 13  Medicine_Category      50000 non-null  object 
 14  Dosage_Form            50000 non-null  object 
 15  St

In [2]:
df["Transaction_Date"] = pd.to_datetime(df["Transaction_Date"])

df["Expiry_Date"] = pd.to_datetime(df["Expiry_Date"])

In [3]:
print("Transaction date:")
print("Earliest:", df["Transaction_Date"].min())
print("Latest:", df["Transaction_Date"].max())

print("\nExpiry date:")
print("Earliest:", df["Expiry_Date"].min())
print("Latest:", df["Expiry_Date"].max())

Transaction date:
Earliest: 2022-01-01 00:00:00
Latest: 2023-06-30 00:00:00

Expiry date:
Earliest: 2022-06-30 00:00:00
Latest: 2025-12-16 00:00:00


In [4]:
expired_sales = df[df["Expiry_Date"] < df["Transaction_Date"]]

print("Transactions sold after expiry:", len(expired_sales))

Transactions sold after expiry: 0


In [5]:
# Missing values 
missing = pd.DataFrame({
    "Missing_Count": df.isna().sum(),
})

missing

,Missing_Count
Transaction_ID,0
Transaction_Date,0
Branch_ID,0
Branch_Name,0
District,0
Province,0
Payment_Method,0
Customer_Gender,0
Customer_Age,0
Customer_City,0


In [6]:
# Number of duplicated rows
print("Duplicated rows:", df.duplicated().sum())

Duplicated rows: 0


In [7]:
# Number of unique values per column
df.nunique().sort_values()

categorical_columns = [
    "Branch_Name",
    "District",
    "Province",
    "Payment_Method",
    "Customer_Gender",
    "Customer_City",
    "Medicine_Category",
    "Dosage_Form",
    "Supplier_Name",
    "Prescription_Required"
]

for col in categorical_columns:
    print(f"\n--- {col} ---")
    print(df[col].value_counts(dropna=False))


--- Branch_Name ---
Branch_Name
Pharmacy - Colombo Fort    7247
Pharmacy - Kurunegala      7236
Pharmacy - Ratnapura       7201
Pharmacy - Jaffna          7160
Pharmacy - Galle           7104
Pharmacy - Negombo         7041
Pharmacy - Kandy City      7011
Name: count, dtype: int64

--- District ---
District
Colombo       7247
Kurunegala    7236
Ratnapura     7201
Jaffna        7160
Galle         7104
Gampaha       7041
Kandy         7011
Name: count, dtype: int64

--- Province ---
Province
Western          14288
North Western     7236
Sabaragamuwa      7201
Northern          7160
Southern          7104
Central           7011
Name: count, dtype: int64

--- Payment_Method ---
Payment_Method
Cash         22652
Card         19886
Insurance     7462
Name: count, dtype: int64

--- Customer_Gender ---
Customer_Gender
Female    25877
Male      24123
Name: count, dtype: int64

--- Customer_City ---
Customer_City
Kurunegala    6433
Kandy         6291
Galle         6258
Colombo       6250
Jaffna

In [8]:
df.describe()

,Transaction_ID,Transaction_Date,Branch_ID,Customer_Age,Medicine_ID,Strength_mg,Pack_Size,Expiry_Date,Quantity,Unit_Price_LKR,Discount_Rate,Line_Total_LKR,Year,Month
count,50000.000000,50000,50000.000000,50000.000000,50000.000000,18764.000000,50000.000000,50000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.00000
mean,25000.500000,2022-09-30 02:25:10.848000,4.001940,48.915940,74.543920,350.273129,35.769760,2024-03-23 08:15:57.888000,3.000780,1878.785603,0.074981,5205.639957,2022.331260,5.51626
min,1.000000,2022-01-01 00:00:00,1.000000,18.000000,1.000000,50.000000,10.000000,2022-06-30 00:00:00,1.000000,26.240000,0.000000,29.340000,2022.000000,1.00000
25%,12500.750000,2022-05-17 00:00:00,2.000000,33.000000,37.000000,100.000000,20.000000,2023-09-11 00:00:00,2.000000,1031.060000,0.040000,2106.485000,2022.000000,3.00000
50%,25000.500000,2022-09-30 00:00:00,4.000000,49.000000,74.000000,250.000000,28.000000,2024-03-23 00:00:00,3.000000,1855.200000,0.070000,4154.685000,2022.000000,5.00000
75%,37500.250000,2023-02-13 00:00:00,6.000000,65.000000,112.000000,500.000000,60.000000,2024-10-04 00:00:00,4.000000,2755.572500,0.110000,7568.817500,2023.000000,8.00000
max,50000.000000,2023-06-30 00:00:00,7.000000,80.000000,148.000000,1000.000000,100.000000,2025-12-16 00:00:00,5.000000,4499.410000,0.150000,21933.450000,2023.000000,12.00000
std,14433.901067,NaN,2.004319,18.159583,42.797969,308.811764,28.103262,NaN,1.412112,1016.935615,0.043655,3960.828905,0.470671,3.30629


In [9]:
numeric_columns = [
    "Customer_Age",
    "Quantity",
    "Unit_Price_LKR",
    "Discount_Rate",
    "Line_Total_LKR"
]

for col in numeric_columns:
    print(f"\n--- {col} ---")
    print("Min:", df[col].min())
    print("Max:", df[col].max())
    print("Negative:", (df[col] < 0).sum())
    print("Zero:", (df[col] == 0).sum())


--- Customer_Age ---
Min: 18
Max: 80
Negative: 0
Zero: 0

--- Quantity ---
Min: 1
Max: 5
Negative: 0
Zero: 0

--- Unit_Price_LKR ---
Min: 26.24
Max: 4499.41
Negative: 0
Zero: 0

--- Discount_Rate ---
Min: 0.0
Max: 0.15
Negative: 0
Zero: 1685

--- Line_Total_LKR ---
Min: 29.34
Max: 21933.45
Negative: 0
Zero: 0


In [10]:
df["Calculated_Total"] = (
    df["Quantity"]
    * df["Unit_Price_LKR"]
    * (1 - df["Discount_Rate"])
).round(2)

df["Total_Difference"] = (
    df["Line_Total_LKR"] - df["Calculated_Total"]
).abs()

print(
    "Incorrect line totals:",
    (df["Total_Difference"] > 0.01).sum()
)

Incorrect line totals: 0


In [11]:
# Check consistency between IDs and their related fields
checks = {
    "Medicine_ID": ["Medicine_Name", "Medicine_Category", "Supplier_Name"],
    "Branch_ID": ["Branch_Name", "District", "Province"]
}

for id_col, related_cols in checks.items():
    for col in related_cols:
        inconsistent = df.groupby(id_col)[col].nunique()
        inconsistent = inconsistent[inconsistent > 1]

        print(f"{id_col} → {col}: {len(inconsistent)} inconsistencies")

Medicine_ID → Medicine_Name: 0 inconsistencies
Medicine_ID → Medicine_Category: 0 inconsistencies
Medicine_ID → Supplier_Name: 0 inconsistencies
Branch_ID → Branch_Name: 0 inconsistencies
Branch_ID → District: 0 inconsistencies
Branch_ID → Province: 0 inconsistencies


In [12]:
# Calculate remaining shelf life at time of sale
df["Days_To_Expiry"] = (
    df["Expiry_Date"] - df["Transaction_Date"]
).dt.days

df["Days_To_Expiry"].describe()

count    50000.00000
mean       540.24360
std        215.85397
min        180.00000
25%        360.00000
50%        540.00000
75%        720.00000
max        900.00000
Name: Days_To_Expiry, dtype: float64

In [13]:
# See which dosage forms have missing Strength_mg
df[df["Strength_mg"].isna()]["Dosage_Form"].value_counts()

Dosage_Form
Cream       7982
Sachet      6320
Inhaler     5814
Ointment    5741
Syrup       5379
Name: count, dtype: int64

In [14]:
# Check if Strength_mg ever have a value for those same dosage forms
df.groupby("Dosage_Form")["Strength_mg"].agg(
    Total="size",
    Missing=lambda x: x.isna().sum(),
    Present=lambda x: x.notna().sum()
)

,Total,Missing,Present
Dosage_Form,,,
Capsule,8103,0,8103
Cream,7982,7982,0
Drops,5045,0,5045
Inhaler,5814,5814,0
Ointment,5741,5741,0
Sachet,6320,6320,0
Syrup,5379,5379,0
Tablet,5616,0,5616


Capsule, Drops, Tablet → always have Strength_mg
Cream, Inhaler, Ointment, Sachet, Syrup → always missing it

In [15]:
# Check customer age range
print("Minimum age:", df["Customer_Age"].min())
print("Maximum age:", df["Customer_Age"].max())



Minimum age: 18
Maximum age: 80


In [16]:
# Check for invalid discount rates
invalid_discount = df[
    (df["Discount_Rate"] < 0) |
    (df["Discount_Rate"] > 1)
]

print("Invalid discount rates:", len(invalid_discount))

Invalid discount rates: 0


In [17]:
# Check invalid quantity or price
invalid_sales = df[
    (df["Quantity"] <= 0) |
    (df["Unit_Price_LKR"] <= 0)
]

print("Invalid quantity/price records:", len(invalid_sales))

Invalid quantity/price records: 0


In [18]:
# Create a cleaned working copy
clean_df = df.copy()
clean_df = clean_df.drop(
    columns=["Calculated_Total", "Total_Difference"],
    errors="ignore"
)

In [19]:
# Standardise column names
clean_df.columns = (
    clean_df.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_")
)

clean_df.columns

Index(['transaction_id', 'transaction_date', 'branch_id', 'branch_name',
       'district', 'province', 'payment_method', 'customer_gender',
       'customer_age', 'customer_city', 'medicine_id', 'medicine_name',
       'generic_name', 'medicine_category', 'dosage_form', 'strength_mg',
       'pack_size', 'supplier_name', 'prescription_required', 'batch_no',
       'expiry_date', 'quantity', 'unit_price_lkr', 'discount_rate',
       'line_total_lkr', 'year', 'month', 'days_to_expiry'],
      dtype='object')

In [20]:
# Remove leading/trailing spaces from all text columns
text_cols = clean_df.select_dtypes(include="object").columns

for col in text_cols:
    clean_df[col] = clean_df[col].str.strip()
clean_df[text_cols].head()


,branch_name,district,province,payment_method,customer_gender,customer_city,medicine_name,generic_name,medicine_category,dosage_form,supplier_name,prescription_required,batch_no
0,Pharmacy - Negombo,Gampaha,Western,Card,Female,Matara,Cipro Vitamin C 100mg Capsule,Vitamin C,Vitamin,Capsule,Ceylon Pharma Traders,No,LK6929-39
1,Pharmacy - Kandy City,Kandy,Central,Card,Female,Kandy,Zithro Zinc Cream 15g,Zinc,Vitamin,Cream,Astron Pharma Supply,No,LK2282-55
2,Pharmacy - Ratnapura,Ratnapura,Sabaragamuwa,Card,Female,Matara,Omep Vitamin D3 50mg Tablet,Vitamin D3,Vitamin,Tablet,Softlogic Pharma Distribution,No,LK1809-97
3,Pharmacy - Kurunegala,Kurunegala,North Western,Cash,Female,Kandy,Gluco Diclofenac Syrup 100ml,Diclofenac,Painkiller,Syrup,Union Chemists Supply,No,LK9152-14
4,Pharmacy - Kurunegala,Kurunegala,North Western,Card,Female,Matara,Cipro ORS 1000mg Tablet,ORS,Wellness,Tablet,Softlogic Pharma Distribution,No,LK6182-19


In [21]:
clean_df["transaction_date"] = pd.to_datetime(
    clean_df["transaction_date"]
)

clean_df["expiry_date"] = pd.to_datetime(
    clean_df["expiry_date"]
)

In [22]:
clean_df = clean_df.rename(
    columns={"line_total_lkr": "revenue_lkr"}
)


print("Rows:", clean_df.shape[0])
print("Columns:", clean_df.shape[1])

print("\nDuplicate rows:", clean_df.duplicated().sum())

print("\nMissing values:")
print(clean_df.isna().sum()[clean_df.isna().sum() > 0])

print("\nData types:")
clean_df.info()

Rows: 50000
Columns: 28

Duplicate rows: 0

Missing values:
strength_mg    31236
dtype: int64

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 28 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   transaction_id         50000 non-null  int64         
 1   transaction_date       50000 non-null  datetime64[ns]
 2   branch_id              50000 non-null  int64         
 3   branch_name            50000 non-null  object        
 4   district               50000 non-null  object        
 5   province               50000 non-null  object        
 6   payment_method         50000 non-null  object        
 7   customer_gender        50000 non-null  object        
 8   customer_age           50000 non-null  int64         
 9   customer_city          50000 non-null  object        
 10  medicine_id            50000 non-null  int64         
 11  medicine_nam

In [23]:
# Time features
clean_df["year"] = clean_df["transaction_date"].dt.year
clean_df["quarter"] = clean_df["transaction_date"].dt.quarter
clean_df["month"] = clean_df["transaction_date"].dt.month
clean_df["month_name"] = clean_df["transaction_date"].dt.month_name()
clean_df["year_month"] = clean_df["transaction_date"].dt.to_period("M").astype(str)
clean_df["day_of_week"] = clean_df["transaction_date"].dt.day_name()

clean_df[[
    "transaction_date",
    "year",
    "quarter",
    "month",
    "month_name",
    "year_month",
    "day_of_week"
]].head()

,transaction_date,year,quarter,month,month_name,year_month,day_of_week
0,2023-06-22,2023,2,6,June,2023-06,Thursday
1,2023-03-27,2023,1,3,March,2023-03,Monday
2,2022-02-15,2022,1,2,February,2022-02,Tuesday
3,2023-03-10,2023,1,3,March,2023-03,Friday
4,2022-12-04,2022,4,12,December,2022-12,Sunday


In [24]:
# Sales before discount
clean_df["gross_sales_lkr"] = (
    clean_df["quantity"] * clean_df["unit_price_lkr"]
)

# Discount amount
clean_df["discount_amount_lkr"] = (
    clean_df["gross_sales_lkr"] - clean_df["revenue_lkr"]
)

clean_df[[
    "quantity",
    "unit_price_lkr",
    "discount_rate",
    "gross_sales_lkr",
    "discount_amount_lkr",
    "revenue_lkr"
]].head()

,quantity,unit_price_lkr,discount_rate,gross_sales_lkr,discount_amount_lkr,revenue_lkr
0,4,1936.52,0.12,7746.08,929.53,6816.55
1,2,2335.56,0.05,4671.12,233.56,4437.56
2,2,2781.10,0.08,5562.20,444.98,5117.22
3,5,1829.65,0.13,9148.25,1189.27,7958.98
4,3,1329.85,0.13,3989.55,518.64,3470.91


In [25]:
age_bins = [0, 18, 30, 45, 60, float("inf")]

age_labels = [
    "Under 18",
    "18-29",
    "30-44",
    "45-59",
    "60+"
]

clean_df["age_group"] = pd.cut(
    clean_df["customer_age"],
    bins=age_bins,
    labels=age_labels,
    right=False
)

clean_df["age_group"].value_counts().sort_index()

age_group
Under 18        0
18-29        9536
30-44       12036
45-59       11898
60+         16530
Name: count, dtype: int64

In [26]:
clean_df["prescription_required"].value_counts()

prescription_required
No     27455
Yes    22545
Name: count, dtype: int64

In [27]:
clean_df["days_to_expiry"] = (
    clean_df["expiry_date"] -
    clean_df["transaction_date"]
).dt.days

In [28]:
clean_df.head()

print("Rows:", clean_df.shape[0])
print("Columns:", clean_df.shape[1])

Rows: 50000
Columns: 35


#EDA

1. OVERALL SALE PERFORMANCE

In [29]:
# Overall sales KPIs
total_revenue = clean_df["revenue_lkr"].sum()
total_units = clean_df["quantity"].sum()
total_transactions = clean_df["transaction_id"].nunique()
avg_transaction_value = total_revenue / total_transactions

print(f"Total Revenue: LKR {total_revenue:,.2f}")
print(f"Units Sold: {total_units:,}")
print(f"Transactions: {total_transactions:,}")
print(f"Average Transaction Value: LKR {avg_transaction_value:,.2f}")

#Revenue by month
monthly_sales = (
    clean_df
    .groupby("year_month")
    .agg(
        revenue=("revenue_lkr", "sum"),
        units_sold=("quantity", "sum"),
        transactions=("transaction_id", "nunique")
    )
    .reset_index()
)

monthly_sales


#Add month-to-month growth
monthly_sales["revenue_growth_pct"] = (
    monthly_sales["revenue"]
    .pct_change() * 100
)

monthly_sales["revenue_growth_pct"] = (
    monthly_sales["revenue_growth_pct"].round(2)
)

monthly_sales

Total Revenue: LKR 260,281,997.87
Units Sold: 150,039
Transactions: 50,000
Average Transaction Value: LKR 5,205.64


,year_month,revenue,units_sold,transactions,revenue_growth_pct
0,2022-01,14327243.74,8592,2878,NaN
1,2022-02,13220670.52,7790,2537,-7.72
2,2022-03,14261010.91,8609,2857,7.87
3,2022-04,13150954.32,7956,2690,-7.78
4,2022-05,14789329.05,8678,2885,12.46
5,2022-06,14663036.42,8487,2819,-0.85
6,2022-07,14457942.64,8550,2826,-1.40
7,2022-08,14410928.70,8392,2818,-0.33
8,2022-09,13604041.58,8079,2729,-5.60
9,2022-10,14653091.45,8515,2844,7.71


In [30]:
best_month = monthly_sales.loc[
    monthly_sales["revenue"].idxmax()
]

worst_month = monthly_sales.loc[
    monthly_sales["revenue"].idxmin()
]

print("Highest revenue month:")
print(best_month)

print("\nLowest revenue month:")
print(worst_month)

Highest revenue month:
year_month                2023-05
revenue               15911181.37
units_sold                   8581
transactions                 2855
revenue_growth_pct           9.71
Name: 16, dtype: object

Lowest revenue month:
year_month                2022-04
revenue               13150954.32
units_sold                   7956
transactions                 2690
revenue_growth_pct          -7.78
Name: 3, dtype: object


In [31]:
#Quarterly Sales Performance
quarterly_sales = (
    clean_df
    .groupby(["year", "quarter"])
    .agg(
        revenue=("revenue_lkr", "sum"),
        units_sold=("quantity", "sum"),
        transactions=("transaction_id", "nunique")
    )
    .reset_index()
)

quarterly_sales

#quarter-over-quarter revenue growth
quarterly_sales["qoq_growth_pct"] = (
    quarterly_sales["revenue"]
    .pct_change() * 100
).round(2)

quarterly_sales



,year,quarter,revenue,units_sold,transactions,qoq_growth_pct
0,2022,1,41808925.17,24991,8272,NaN
1,2022,2,42603319.79,25121,8394,1.90
2,2022,3,42472912.92,25021,8373,-0.31
3,2022,4,43534016.38,25221,8398,2.50
4,2023,1,44599912.47,24877,8289,2.45
5,2023,2,45262911.14,24808,8274,1.49


In [32]:
#Year-over-Year comparison
#compare the same six-month period: Jan–Jun 2022 vs Jan–Jun 2023.
yoy_sales = (
    clean_df[clean_df["month"] <= 6]
    .groupby("year")
    .agg(
        revenue=("revenue_lkr", "sum"),
        units_sold=("quantity", "sum"),
        transactions=("transaction_id", "nunique")
    )
    .reset_index()
)

yoy_sales

yoy_sales["revenue_growth_pct"] = (
    yoy_sales["revenue"].pct_change() * 100
).round(2)

yoy_sales["units_growth_pct"] = (
    yoy_sales["units_sold"].pct_change() * 100
).round(2)

yoy_sales["transaction_growth_pct"] = (
    yoy_sales["transactions"].pct_change() * 100
).round(2)

yoy_sales

,year,revenue,units_sold,transactions,revenue_growth_pct,units_growth_pct,transaction_growth_pct
0,2022,84412244.96,50112,16666,NaN,NaN,NaN
1,2023,89862823.61,49685,16563,6.46,-0.85,-0.62


In [33]:
#average selling value

yoy_sales["revenue_per_unit"] = (
    yoy_sales["revenue"] / yoy_sales["units_sold"]
)

yoy_sales["avg_transaction_value"] = (
    yoy_sales["revenue"] / yoy_sales["transactions"]
)

yoy_sales[[
    "year",
    "revenue_per_unit",
    "avg_transaction_value"
]].round(2)

,year,revenue_per_unit,avg_transaction_value
0,2022,1684.47,5064.94
1,2023,1808.65,5425.52


In [34]:
#Compare average unit price
price_comparison = (
    clean_df[clean_df["month"] <= 6]
    .groupby("year")
    .agg(
        avg_unit_price=("unit_price_lkr", "mean"),
        avg_discount=("discount_rate", "mean")
    )
    .reset_index()
)

price_comparison

,year,avg_unit_price,avg_discount
0,2022,1821.632923,0.074972
1,2023,1959.102408,0.075232


In [35]:
medicine_price = (
    clean_df[clean_df["month"] <= 6]
    .groupby(["medicine_id", "medicine_name", "year"])
    .agg(
        avg_price=("unit_price_lkr", "mean"),
        units_sold=("quantity", "sum")
    )
    .reset_index()
)

medicine_price.head()

,medicine_id,medicine_name,year,avg_price,units_sold
0,1,Gluco Paracetamol 50mg Capsule,2022,1346.971290,294
1,1,Gluco Paracetamol 50mg Capsule,2023,1435.890315,364
2,2,Vita Paracetamol Inhaler,2022,3192.412419,364
3,2,Vita Paracetamol Inhaler,2023,3429.108148,314
4,3,Aller Paracetamol Inhaler,2022,1064.450280,465


In [36]:
#Compare prices of the same medicines

price_change = medicine_price.pivot(
    index=["medicine_id", "medicine_name"],
    columns="year",
    values="avg_price"
).reset_index()

price_change["price_change_pct"] = (
    (price_change[2023] - price_change[2022])
    / price_change[2022]
    * 100
).round(2)

price_change.sort_values(
    "price_change_pct",
    ascending=False
).head(10)

year,medicine_id,medicine_name,2022,2023,price_change_pct
3,4,Amoxil Paracetamol 1000mg Capsule,2628.819677,2867.368534,9.07
54,55,Amoxil Aspirin (EC) 125mg Drops,2514.699817,2724.047417,8.32
101,102,Omep Hydrocortisone Sachet,3281.671583,3549.416961,8.16
48,49,Vita Bisoprolol 75mg Drops,1946.998704,2102.959580,8.01
127,128,Amoxil Acyclovir Syrup 100ml,843.545405,910.425393,7.93
121,122,Aller Fluconazole Ointment 15g,1141.081089,1229.901217,7.78
22,23,Cardio Ciprofloxacin Cream 15g,3113.128712,3352.034632,7.67
76,77,Cipro ORS 1000mg Tablet,1254.684947,1350.823077,7.66
114,115,Cardio Co-trimoxazole 75mg Capsule,1661.370926,1787.246500,7.58
134,135,Aller Prednisolone 500mg Capsule,2117.479459,2277.620185,7.56


In [37]:
price_change["price_change_pct"].describe()

count    148.000000
mean       5.993311
std        1.136541
min        1.850000
25%        5.170000
50%        6.090000
75%        6.795000
max        9.070000
Name: price_change_pct, dtype: float64

In [38]:
print("Price increased:",
      (price_change["price_change_pct"] > 0).sum())

print("Price decreased:",
      (price_change["price_change_pct"] < 0).sum())

print("No change:",
      (price_change["price_change_pct"] == 0).sum())

Price increased: 148
Price decreased: 0
No change: 0


So the 2023 revenue increase is likely strongly related to higher medicine prices, rather than increased demand:
Revenue: +6.46%
Units sold: −0.85%
Transactions: −0.62%
All 148 medicines increased in price
Average unit price: +7.55%
Discounts: essentially unchanged

In [39]:
price_change["price_change_pct"].describe().round(2)

count    148.00
mean       5.99
std        1.14
min        1.85
25%        5.17
50%        6.09
75%        6.80
max        9.07
Name: price_change_pct, dtype: float64

Jan–Jun 2023 revenue grew 6.46% compared with Jan–Jun 2022, despite slight declines in units sold and transactions. The growth was primarily price-driven, with all 148 medicines increasing in price and the median medicine price rising 6.09%.

In [40]:
#category performance
category_performance = (
    clean_df
    .groupby("medicine_category")
    .agg(
        revenue=("revenue_lkr", "sum"),
        units_sold=("quantity", "sum"),
        transactions=("transaction_id", "nunique"),
        medicines=("medicine_id", "nunique")
    )
    .reset_index()
    .sort_values("revenue", ascending=False)
)

category_performance

,medicine_category,revenue,units_sold,transactions,medicines
4,Cardiac,36729665.88,19401,6499,19
1,Antibiotic,33956397.96,23116,7632,22
13,Vitamin,29562797.26,14668,4873,15
11,Painkiller,24206932.15,15090,5008,15
5,Cough & Cold,23898672.64,12158,4088,12
6,Dermatology,20539635.42,9994,3363,10
8,First Aid,16685842.93,8172,2707,8
7,Diabetes,15640762.17,10266,3377,10
9,Gastro,14521448.14,10230,3436,10
10,Hormonal,10289525.97,7085,2363,7


In [41]:
#compare revenue rank vs unit-sales rank. 
#helps separate categories that make lots of money because they're expensive from categories with genuinely high demand.
category_performance["revenue_rank"] = (
    category_performance["revenue"]
    .rank(ascending=False, method="dense")
    .astype(int)
)

category_performance["units_rank"] = (
    category_performance["units_sold"]
    .rank(ascending=False, method="dense")
    .astype(int)
)

category_performance[
    [
        "medicine_category",
        "revenue",
        "units_sold",
        "revenue_rank",
        "units_rank"
    ]
].sort_values("revenue_rank")

,medicine_category,revenue,units_sold,revenue_rank,units_rank
4,Cardiac,36729665.88,19401,1,2
1,Antibiotic,33956397.96,23116,2,1
13,Vitamin,29562797.26,14668,3,4
11,Painkiller,24206932.15,15090,4,3
5,Cough & Cold,23898672.64,12158,5,5
6,Dermatology,20539635.42,9994,6,8
8,First Aid,16685842.93,8172,7,9
7,Diabetes,15640762.17,10266,8,6
9,Gastro,14521448.14,10230,9,7
10,Hormonal,10289525.97,7085,10,10


In [42]:
revenue_median = category_performance["revenue"].median()
units_median = category_performance["units_sold"].median()

def classify_category(row):
    if row["revenue"] >= revenue_median and row["units_sold"] >= units_median:
        return "Core Category"
    
    elif row["revenue"] >= revenue_median and row["units_sold"] < units_median:
        return "High Value"
    
    elif row["revenue"] < revenue_median and row["units_sold"] >= units_median:
        return "High Demand, Lower Value"
    
    else:
        return "Lower Priority"

category_performance["category_class"] = (
    category_performance.apply(classify_category, axis=1)
)

category_performance[
    [
        "medicine_category",
        "revenue",
        "units_sold",
        "category_class"
    ]
].sort_values(
    ["category_class", "revenue"],
    ascending=[True, False]
)

,medicine_category,revenue,units_sold,category_class
4,Cardiac,36729665.88,19401,Core Category
1,Antibiotic,33956397.96,23116,Core Category
13,Vitamin,29562797.26,14668,Core Category
11,Painkiller,24206932.15,15090,Core Category
5,Cough & Cold,23898672.64,12158,Core Category
6,Dermatology,20539635.42,9994,Core Category
7,Diabetes,15640762.17,10266,Core Category
9,Gastro,14521448.14,10230,"High Demand, Lower Value"
8,First Aid,16685842.93,8172,High Value
10,Hormonal,10289525.97,7085,Lower Priority


In [43]:
medicine_performance = (
    clean_df
    .groupby(["medicine_id", "medicine_name", "medicine_category"])
    .agg(
        revenue=("revenue_lkr", "sum"),
        units_sold=("quantity", "sum"),
        transactions=("transaction_id", "nunique"),
        avg_unit_price=("unit_price_lkr", "mean")
    )
    .reset_index()
)


#contribution to total sales
medicine_performance["revenue_share_pct"] = (
    medicine_performance["revenue"]
    / medicine_performance["revenue"].sum()
    * 100
).round(2)

medicine_performance["unit_share_pct"] = (
    medicine_performance["units_sold"]
    / medicine_performance["units_sold"].sum()
    * 100
).round(2)

medicine_performance.sort_values(
    "revenue",
    ascending=False
).head(10)

,medicine_id,medicine_name,medicine_category,revenue,units_sold,transactions,avg_unit_price,revenue_share_pct,unit_share_pct
72,73,Vita Domperidone Cream 15g,Gastro,3659641.31,1146,378,3439.336720,1.41,0.76
22,23,Cardio Ciprofloxacin Cream 15g,Antibiotic,3301511.36,1104,366,3238.237377,1.27,0.74
85,86,Cipro Iron + Folic Sachet,Vitamin,3296558.10,996,324,3579.524074,1.27,0.66
10,11,Zithro Diclofenac Ointment 15g,Painkiller,3239726.98,1008,344,3456.057965,1.24,0.67
62,63,Amoxil Salbutamol Ointment 15g,Respiratory,3154746.28,1047,346,3254.171012,1.21,0.70
136,137,Gluco Dextromethorphan Inhaler,Cough & Cold,3153287.78,1030,344,3302.696163,1.21,0.69
45,46,Amoxil Amlodipine 400mg Tablet,Cardiac,3126833.12,1006,344,3358.139070,1.20,0.67
101,102,Omep Hydrocortisone Sachet,Dermatology,3102246.87,981,323,3408.642693,1.19,0.65
41,42,Vita Losartan Cream 15g,Cardiac,3092670.35,1092,360,3064.090611,1.19,0.73
36,37,Aller Atorvastatin 1000mg Capsule,Cardiac,3089522.52,1044,343,3207.663294,1.19,0.70


In [44]:
contribution_analysis = medicine_performance.sort_values(
    "revenue",
    ascending=False
).copy()

contribution_analysis["cumulative_revenue_pct"] = (
    contribution_analysis["revenue"].cumsum()
    / contribution_analysis["revenue"].sum()
    * 100
)

def classify_contribution(cumulative_pct):
    if cumulative_pct <= 80:
        return "top"
    elif cumulative_pct <= 95:
        return "middle"
    else:
        return "bottom"

contribution_analysis["contribution_class"] = (
    contribution_analysis["cumulative_revenue_pct"]
    .apply(classify_contribution)
)


contribution_summary = (
    contribution_analysis
    .groupby("contribution_class")
    .agg(
        medicines=("medicine_id", "count"),
        revenue=("revenue", "sum"),
        units_sold=("units_sold", "sum")
    )
    .reset_index()
)

contribution_summary["revenue_share_pct"] = (
    contribution_summary["revenue"]
    / contribution_summary["revenue"].sum()
    * 100
).round(2)

contribution_summary

,contribution_class,medicines,revenue,units_sold,revenue_share_pct
0,bottom,28,1.337920e+07,28657,5.14
1,middle,35,3.953527e+07,35758,15.19
2,top,85,2.073675e+08,85624,79.67


In [45]:
contribution_summary["medicine_share_pct"] = (
    contribution_summary["medicines"]
    / contribution_summary["medicines"].sum()
    * 100
).round(2)

contribution_summary

,contribution_class,medicines,revenue,units_sold,revenue_share_pct,medicine_share_pct
0,bottom,28,1.337920e+07,28657,5.14,18.92
1,middle,35,3.953527e+07,35758,15.19,23.65
2,top,85,2.073675e+08,85624,79.67,57.43


In [46]:
#Fast vs slow movers
medicine_velocity = (
    clean_df
    .groupby(["medicine_id", "medicine_name"])
    .agg(
        total_units=("quantity", "sum"),
        months_sold=("year_month", "nunique")
    )
    .reset_index()
)

medicine_velocity["avg_units_per_month"] = (
    medicine_velocity["total_units"]
    / medicine_velocity["months_sold"]
).round(2)

medicine_velocity.head()

,medicine_id,medicine_name,total_units,months_sold,avg_units_per_month
0,1,Gluco Paracetamol 50mg Capsule,1042,18,57.89
1,2,Vita Paracetamol Inhaler,978,18,54.33
2,3,Aller Paracetamol Inhaler,1069,18,59.39
3,4,Amoxil Paracetamol 1000mg Capsule,1090,18,60.56
4,5,Aller Paracetamol Inhaler,1015,18,56.39


In [47]:
medicine_velocity["avg_units_per_month"].describe().round(2)




count    148.00
mean      56.32
std        3.39
min       47.72
25%       54.14
50%       56.53
75%       58.44
max       64.06
Name: avg_units_per_month, dtype: float64

In [48]:
q1 = medicine_velocity["avg_units_per_month"].quantile(0.25)
q3 = medicine_velocity["avg_units_per_month"].quantile(0.75)

def classify_demand(velocity):
    if velocity >= q3:
        return "High Demand"
    elif velocity <= q1:
        return "Low Demand"
    else:
        return "Moderate Demand"

medicine_velocity["demand_class"] = (
    medicine_velocity["avg_units_per_month"]
    .apply(classify_demand)
)

medicine_velocity["demand_class"].value_counts()

#combine with contribution
medicine_analysis = contribution_analysis.merge(
    medicine_velocity[
        ["medicine_id", "avg_units_per_month", "demand_class"]
    ],
    on="medicine_id",
    how="left"
)

pd.crosstab(
    medicine_analysis["contribution_class"],
    medicine_analysis["demand_class"]
)

demand_class,High Demand,Low Demand,Moderate Demand
contribution_class,,,
bottom,7,5,16
middle,12,9,14
top,19,23,43


In [49]:
#Branch performance
branch_performance = (
    clean_df
    .groupby(["branch_id", "branch_name", "district", "province"])
    .agg(
        revenue=("revenue_lkr", "sum"),
        units_sold=("quantity", "sum"),
        transactions=("transaction_id", "nunique"),
        medicines_sold=("medicine_id", "nunique")
    )
    .reset_index()
)

branch_performance["avg_transaction_value"] = (
    branch_performance["revenue"]
    / branch_performance["transactions"]
).round(2)

branch_performance["revenue_share_pct"] = (
    branch_performance["revenue"]
    / branch_performance["revenue"].sum()
    * 100
).round(2)

branch_performance.sort_values(
    "revenue",
    ascending=False
)

,branch_id,branch_name,district,province,revenue,units_sold,transactions,medicines_sold,avg_transaction_value,revenue_share_pct
0,1,Pharmacy - Colombo Fort,Colombo,Western,38041984.10,21754,7247,148,5249.34,14.62
6,7,Pharmacy - Ratnapura,Ratnapura,Sabaragamuwa,37860964.66,21729,7201,148,5257.74,14.55
3,4,Pharmacy - Kurunegala,Kurunegala,North Western,37561479.29,21735,7236,148,5190.92,14.43
2,3,Pharmacy - Galle,Galle,Southern,37075300.38,21310,7104,148,5218.93,14.24
4,5,Pharmacy - Negombo,Gampaha,Western,37042941.41,21247,7041,148,5261.03,14.23
5,6,Pharmacy - Jaffna,Jaffna,Northern,36672577.19,21421,7160,148,5121.87,14.09
1,2,Pharmacy - Kandy City,Kandy,Central,36026750.84,20843,7011,148,5138.60,13.84


In [50]:
#branch growth


branch_yoy = (
    clean_df[clean_df["month"] <= 6]
    .groupby(["branch_id", "branch_name", "year"])
    .agg(
        revenue=("revenue_lkr", "sum"),
        units_sold=("quantity", "sum"),
        transactions=("transaction_id", "nunique")
    )
    .reset_index()
)

branch_growth = branch_yoy.pivot(
    index=["branch_id", "branch_name"],
    columns="year",
    values="revenue"
).reset_index()

branch_growth["revenue_growth_pct"] = (
    (branch_growth[2023] - branch_growth[2022])
    / branch_growth[2022]
    * 100
).round(2)

branch_growth.sort_values(
    "revenue_growth_pct",
    ascending=False
)

year,branch_id,branch_name,2022,2023,revenue_growth_pct
0,1,Pharmacy - Colombo Fort,11978111.90,13107836.49,9.43
2,3,Pharmacy - Galle,11865417.64,12982717.27,9.42
5,6,Pharmacy - Jaffna,11791926.91,12787651.81,8.44
4,5,Pharmacy - Negombo,12032555.01,12902134.37,7.23
3,4,Pharmacy - Kurunegala,12049521.74,12723256.02,5.59
1,2,Pharmacy - Kandy City,11924944.05,12569983.41,5.41
6,7,Pharmacy - Ratnapura,12769767.71,12789244.24,0.15


In [51]:
#supplier performance

supplier_performance = (
    clean_df
    .groupby("supplier_name")
    .agg(
        revenue=("revenue_lkr", "sum"),
        units_sold=("quantity", "sum"),
        transactions=("transaction_id", "nunique"),
        medicines_supplied=("medicine_id", "nunique")
    )
    .reset_index()
)

supplier_performance["revenue_share_pct"] = (
    supplier_performance["revenue"]
    / supplier_performance["revenue"].sum()
    * 100
).round(2)

supplier_performance.sort_values(
    "revenue",
    ascending=False
)

,supplier_name,revenue,units_sold,transactions,medicines_supplied,revenue_share_pct
7,Union Chemists Supply,39149100.71,22252,7415,22,15.04
3,Healthguard Central Supply,35674934.38,16376,5459,16,13.71
1,Ceylon Pharma Traders,35193547.57,16881,5647,17,13.52
2,Ceyoka Healthcare Supply,34780206.71,23028,7701,23,13.36
0,Astron Pharma Supply,34512941.11,21713,7195,21,13.26
4,Hemas Pharma Wholesale,33996152.55,18607,6184,18,13.06
6,Softlogic Pharma Distribution,33647924.42,21534,7130,21,12.93
5,SPC Distribution (LK),13327190.42,9648,3269,10,5.12


In [52]:
#Supplier contribution
supplier_contribution = (
    supplier_performance
    .sort_values("revenue", ascending=False)
    .copy()
)

supplier_contribution["cumulative_revenue_pct"] = (
    supplier_contribution["revenue"].cumsum()
    / supplier_contribution["revenue"].sum()
    * 100
).round(2)

supplier_contribution[
    [
        "supplier_name",
        "medicines_supplied",
        "revenue",
        "revenue_share_pct",
        "cumulative_revenue_pct"
    ]
]

top_3_supplier_share = (
    supplier_contribution
    .head(3)["revenue"]
    .sum()
    / supplier_contribution["revenue"].sum()
    * 100
)

print(f"Top 3 suppliers account for {top_3_supplier_share:.2f}% of revenue")

Top 3 suppliers account for 42.27% of revenue


In [53]:
customer_analysis = (
    clean_df
    .groupby("age_group", observed=True)
    .agg(
        revenue=("revenue_lkr", "sum"),
        units_sold=("quantity", "sum"),
        transactions=("transaction_id", "nunique")
    )
    .reset_index()
)

customer_analysis["avg_transaction_value"] = (
    customer_analysis["revenue"]
    / customer_analysis["transactions"]
).round(2)

customer_analysis["revenue_share_pct"] = (
    customer_analysis["revenue"]
    / customer_analysis["revenue"].sum()
    * 100
).round(2)

customer_analysis

,age_group,revenue,units_sold,transactions,avg_transaction_value,revenue_share_pct
0,18-29,49588622.52,28480,9536,5200.15,19.05
1,30-44,62573780.92,36048,12036,5198.89,24.04
2,45-59,62088945.84,35806,11898,5218.44,23.85
3,60+,86030648.59,49705,16530,5204.52,33.05


In [54]:
#Prescription requirement performance
prescription_analysis = (
    clean_df
    .groupby("prescription_required")
    .agg(
        revenue=("revenue_lkr", "sum"),
        units_sold=("quantity", "sum"),
        transactions=("transaction_id", "nunique"),
        medicines=("medicine_id", "nunique")
    )
    .reset_index()
)

prescription_analysis["revenue_share_pct"] = (
    prescription_analysis["revenue"]
    / prescription_analysis["revenue"].sum()
    * 100
).round(2)

prescription_analysis["avg_transaction_value"] = (
    prescription_analysis["revenue"]
    / prescription_analysis["transactions"]
).round(2)

prescription_analysis

,prescription_required,revenue,units_sold,transactions,medicines,revenue_share_pct,avg_transaction_value
0,No,1.472030e+08,82240,27455,82,56.56,5361.61
1,Yes,1.130790e+08,67799,22545,66,43.44,5015.70


In [55]:
#compare average price, 
#because prescription medicines might generate more revenue simply 
#because they're more expensive
rx_price = (
    clean_df
    .groupby("prescription_required")
    .agg(
        avg_unit_price=("unit_price_lkr", "mean"),
        avg_discount=("discount_rate", "mean")
    )
    .reset_index()
)

rx_price.round(2)

,prescription_required,avg_unit_price,avg_discount
0,No,1936.22,0.08
1,Yes,1808.84,0.07


In [56]:
#Discount analysis

def classify_discount(rate):
    if rate == 0:
        return "No Discount"
    elif rate <= 0.05:
        return "Low (1-5%)"
    elif rate <= 0.10:
        return "Medium (6-10%)"
    else:
        return "High (>10%)"

clean_df["discount_group"] = (
    clean_df["discount_rate"]
    .apply(classify_discount)
)

In [57]:
#summarise performance:
discount_analysis = (
    clean_df
    .groupby("discount_group")
    .agg(
        revenue=("revenue_lkr", "sum"),
        units_sold=("quantity", "sum"),
        transactions=("transaction_id", "nunique"),
        avg_quantity=("quantity", "mean")
    )
    .reset_index()
)

discount_analysis["revenue_share_pct"] = (
    discount_analysis["revenue"]
    / discount_analysis["revenue"].sum()
    * 100
).round(2)

discount_analysis.round(2)

,discount_group,revenue,units_sold,transactions,avg_quantity,revenue_share_pct
0,High (>10%),74548336.64,45537,15137,3.01,28.64
1,Low (1-5%),90615883.87,49910,16681,2.99,34.81
2,Medium (6-10%),85539200.53,49476,16497,3.00,32.86
3,No Discount,9578576.83,5116,1685,3.04,3.68


In [58]:
#How much revenue is being given up through discounts?
discount_impact = clean_df.agg(
    gross_sales=("gross_sales_lkr", "sum"),
    discount_amount=("discount_amount_lkr", "sum"),
    net_revenue=("revenue_lkr", "sum")
)

discount_impact

,gross_sales_lkr,discount_amount_lkr,revenue_lkr
gross_sales,2.814301e+08,NaN,NaN
discount_amount,NaN,21148079.51,NaN
net_revenue,NaN,NaN,2.602820e+08


In [59]:
effective_discount_rate = (
    clean_df["discount_amount_lkr"].sum()
    / clean_df["gross_sales_lkr"].sum()
    * 100
)

print(f"Effective discount rate: {effective_discount_rate:.2f}%")

Effective discount rate: 7.51%


In [60]:
#Monthly seasonality by category

category_monthly = (
    clean_df
    .groupby(["medicine_category", "month", "month_name"])
    .agg(
        units_sold=("quantity", "sum"),
        revenue=("revenue_lkr", "sum")
    )
    .reset_index()
)

category_monthly.head()

,medicine_category,month,month_name,units_sold,revenue
0,Allergy,1,January,740,1022357.75
1,Allergy,2,February,645,950031.93
2,Allergy,3,March,708,984493.88
3,Allergy,4,April,543,796529.31
4,Allergy,5,May,698,1067467.59


In [61]:
#average monthly units
category_avg = (
    category_monthly
    .groupby("medicine_category")["units_sold"]
    .mean()
    .rename("avg_monthly_units")
)

category_monthly = category_monthly.merge(
    category_avg,
    on="medicine_category"
)

category_monthly["demand_index"] = (
    category_monthly["units_sold"]
    /category_monthly["avg_monthly_units"]
    * 100
).round(1)

peak_months = (
    category_monthly
    .loc[
        category_monthly
        .groupby("medicine_category")["demand_index"]
        .idxmax()
    ]
    [["medicine_category", "month_name", "demand_index"]]
    .sort_values("demand_index", ascending=False)
)

peak_months

,medicine_category,month_name,demand_index
144,Respiratory,January,156.8
29,Antifungal,June,156.7
120,Hormonal,January,151.9
170,Wellness,March,149.4
64,Cough & Cold,May,147.4
40,Antiviral,May,146.5
72,Dermatology,January,145.9
0,Allergy,January,144.7
86,Diabetes,March,144.7
132,Painkiller,January,144.2


In [62]:
def classify_priority(row):
    contribution = row["contribution_class"]
    demand = row["demand_class"]

    if contribution == "top" and demand == "High Demand":
        return "Priority Stock"

    elif contribution == "top":
        return "High Value"

    elif demand == "High Demand":
        return "High Demand"

    elif contribution == "bottom" and demand == "Low Demand":
        return "Review"

    else:
        return "Standard"

medicine_analysis["management_priority"] = (
    medicine_analysis.apply(classify_priority, axis=1)
)

In [63]:
medicine_analysis["management_priority"].value_counts()

management_priority
High Value        66
Standard          39
Priority Stock    19
High Demand       19
Review             5
Name: count, dtype: int64

In [64]:
medicine_analysis[
    [
        "medicine_name",
        "medicine_category",
        "revenue",
        "avg_units_per_month",
        "contribution_class",
        "demand_class",
        "management_priority"
    ]
].sort_values(
    ["management_priority", "revenue"],
    ascending=[True, False]
)

,medicine_name,medicine_category,revenue,avg_units_per_month,contribution_class,demand_class,management_priority
88,Panadol Azithromycin Cream 15g,Antibiotic,1376903.27,62.06,middle,High Demand,High Demand
91,Calpol Losartan Ointment 15g,Cardiac,1332960.88,61.17,middle,High Demand,High Demand
94,Aller Domperidone Cream 15g,Gastro,1316974.32,58.61,middle,High Demand,High Demand
96,Panadol Insulin (Human) Inhaler,Diabetes,1238832.19,59.50,middle,High Demand,High Demand
100,Zithro Atorvastatin Inhaler,Cardiac,1148129.36,58.89,middle,High Demand,High Demand
...,...,...,...,...,...,...,...
143,Cardio Domperidone 1000mg Drops,Gastro,115625.05,57.33,bottom,Moderate Demand,Standard
144,Omep Phenylephrine 650mg Capsule,Cough & Cold,75384.87,57.50,bottom,Moderate Demand,Standard
145,Panadol Ciprofloxacin 400mg Tablet,Antibiotic,75219.76,58.33,bottom,Moderate Demand,Standard
146,Amoxil Ibuprofen 250mg Drops,Painkiller,71661.91,55.39,bottom,Moderate Demand,Standard


In [65]:
print("Rows:", clean_df.shape[0])
print("Columns:", clean_df.shape[1])

clean_df.head()

Rows: 50000
Columns: 36


,transaction_id,transaction_date,branch_id,branch_name,district,province,payment_method,customer_gender,customer_age,customer_city,...,month,days_to_expiry,quarter,month_name,year_month,day_of_week,gross_sales_lkr,discount_amount_lkr,age_group,discount_group
0,1,2023-06-22,5,Pharmacy - Negombo,Gampaha,Western,Card,Female,45,Matara,...,6,570,2,June,2023-06,Thursday,7746.08,929.53,45-59,High (>10%)
1,2,2023-03-27,2,Pharmacy - Kandy City,Kandy,Central,Card,Female,75,Kandy,...,3,390,1,March,2023-03,Monday,4671.12,233.56,60+,Low (1-5%)
2,3,2022-02-15,7,Pharmacy - Ratnapura,Ratnapura,Sabaragamuwa,Card,Female,20,Matara,...,2,360,1,February,2022-02,Tuesday,5562.20,444.98,18-29,Medium (6-10%)
3,4,2023-03-10,4,Pharmacy - Kurunegala,Kurunegala,North Western,Cash,Female,47,Kandy,...,3,510,1,March,2023-03,Friday,9148.25,1189.27,45-59,High (>10%)
4,5,2022-12-04,4,Pharmacy - Kurunegala,Kurunegala,North Western,Card,Female,19,Matara,...,12,420,4,December,2022-12,Sunday,3989.55,518.64,18-29,High (>10%)


In [66]:
clean_df.columns.tolist()

['transaction_id',
 'transaction_date',
 'branch_id',
 'branch_name',
 'district',
 'province',
 'payment_method',
 'customer_gender',
 'customer_age',
 'customer_city',
 'medicine_id',
 'medicine_name',
 'generic_name',
 'medicine_category',
 'dosage_form',
 'strength_mg',
 'pack_size',
 'supplier_name',
 'prescription_required',
 'batch_no',
 'expiry_date',
 'quantity',
 'unit_price_lkr',
 'discount_rate',
 'revenue_lkr',
 'year',
 'month',
 'days_to_expiry',
 'quarter',
 'month_name',
 'year_month',
 'day_of_week',
 'gross_sales_lkr',
 'discount_amount_lkr',
 'age_group',
 'discount_group']

In [67]:
clean_df.to_csv(
    "pharmacy_sales_clean.csv",
    index=False
)

In [69]:
dim_branch = (
    clean_df[
        ["branch_id", "branch_name", "district", "province"]
    ]
    .drop_duplicates()
    .sort_values("branch_id")
    .reset_index(drop=True)
)

dim_branch

,branch_id,branch_name,district,province
0,1,Pharmacy - Colombo Fort,Colombo,Western
1,2,Pharmacy - Kandy City,Kandy,Central
2,3,Pharmacy - Galle,Galle,Southern
3,4,Pharmacy - Kurunegala,Kurunegala,North Western
4,5,Pharmacy - Negombo,Gampaha,Western
5,6,Pharmacy - Jaffna,Jaffna,Northern
6,7,Pharmacy - Ratnapura,Ratnapura,Sabaragamuwa


In [70]:
dim_branch.to_csv("dim_branch.csv", index=False)


In [71]:
dim_medicine = (
    clean_df[
        [
            "medicine_id",
            "medicine_name",
            "generic_name",
            "medicine_category",
            "dosage_form",
            "strength_mg",
            "pack_size",
            "supplier_name",
            "prescription_required"
        ]
    ]
    .drop_duplicates()
    .sort_values("medicine_id")
    .reset_index(drop=True)
)

dim_medicine.to_csv("dim_medicine.csv", index=False)

print(dim_medicine.shape)

(148, 9)


In [72]:
dim_medicine_export = dim_medicine.copy()

dim_medicine_export["strength_mg"] = (
    dim_medicine_export["strength_mg"]
    .where(dim_medicine_export["strength_mg"].notna(), "")
)

dim_medicine_export.to_csv(
    "dim_medicine.csv",
    index=False
)

In [73]:
dim_date = pd.DataFrame({
    "date": pd.date_range(
        start=clean_df["transaction_date"].min(),
        end=clean_df["transaction_date"].max()
    )
})

dim_date["year"] = dim_date["date"].dt.year
dim_date["quarter"] = dim_date["date"].dt.quarter
dim_date["month"] = dim_date["date"].dt.month
dim_date["month_name"] = dim_date["date"].dt.month_name()
dim_date["year_month"] = dim_date["date"].dt.to_period("M").astype(str)
dim_date["day_of_week"] = dim_date["date"].dt.day_name()

dim_date.head()

,date,year,quarter,month,month_name,year_month,day_of_week
0,2022-01-01,2022,1,1,January,2022-01,Saturday
1,2022-01-02,2022,1,1,January,2022-01,Sunday
2,2022-01-03,2022,1,1,January,2022-01,Monday
3,2022-01-04,2022,1,1,January,2022-01,Tuesday
4,2022-01-05,2022,1,1,January,2022-01,Wednesday


In [74]:
dim_date.to_csv("dim_date.csv", index=False)

In [75]:
dim_date_export = dim_date.drop(
    columns=["year_month"],
    errors="ignore"
)

dim_date_export.to_csv(
    "dim_date.csv",
    index=False
)

dim_date_export.head()

,date,year,quarter,month,month_name,day_of_week
0,2022-01-01,2022,1,1,January,Saturday
1,2022-01-02,2022,1,1,January,Sunday
2,2022-01-03,2022,1,1,January,Monday
3,2022-01-04,2022,1,1,January,Tuesday
4,2022-01-05,2022,1,1,January,Wednesday


In [76]:
fact_sales = clean_df[
    [
        "transaction_id",
        "transaction_date",
        "branch_id",
        "medicine_id",
        "payment_method",
        "customer_gender",
        "customer_age",
        "customer_city",
        "batch_no",
        "expiry_date",
        "quantity",
        "unit_price_lkr",
        "discount_rate",
        "gross_sales_lkr",
        "discount_amount_lkr",
        "revenue_lkr",
        "days_to_expiry",
        "age_group",
        "discount_group"
    ]
].copy()

fact_sales.head()

fact_sales.to_csv("fact_sales.csv", index=False)

In [77]:
with pd.ExcelWriter("Pharmacy_Analytics.xlsx", engine="openpyxl") as writer:
    
    fact_sales.to_excel(
        writer,
        sheet_name="FactSales",
        index=False
    )
    
    dim_date_export.to_excel(
        writer,
        sheet_name="DimDate",
        index=False
    )
    
    dim_branch.to_excel(
        writer,
        sheet_name="DimBranch",
        index=False
    )
    
    dim_medicine.to_excel(
        writer,
        sheet_name="DimMedicine",
        index=False
    )

print("Pharmacy_Analytics.xlsx created successfully!")

Pharmacy_Analytics.xlsx created successfully!


In [78]:
clean_df["customer_age"].agg(["min", "max"])

min    18
max    80
Name: customer_age, dtype: int64